File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int left = 1, right = n - 2;
        int lmax = height[left - 1];
        int rmax = height[right + 1];
        int result = 0;
        while (left <= right) {
            if (lmax < rmax) {
                result += max(0, lmax - height[left]);
                lmax = max(lmax, height[left]);
                left++;
            } else {
                result += max(0, rmax - height[right]);
                rmax = max(rmax, height[right]);
                right--;
            }
        }

        return result;
    }
};

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l_max, r_max = height[0], height[-1]
        l, r = 1, len(height) - 2
        water = 0
        while l <= r:
            if l_max < r_max:
                water += max(0, l_max - height[l])
                l_max = max(height[l], l_max)
                l += 1
            else:
                water += max(0, r_max - height[r])
                r_max = max(height[r], r_max)
                r -= 1
        
        return water

                


        